From d7ecb7ad728ab643d3aa12cf06fd4332532d2410 Mon Sep 17 00:00:00 2001
From: Hanna Hawa <hannah@marvell.com>
Date: Mon, 27 Jul 2015 13:25:52 +0300
Subject: [PATCH 0281/1240] tools: xdb: added support for XDB scripts to run
 dynamic-spl

- Added support for XDB scripts to run dynamic spl
- Tested on A38x (with spl + u-boot)
- Tested on Palladuim-AP806 (with spl)

Change-Id: Iea92789984344d68ca02086e13be947df3812016
Signed-off-by: Hanna Hawa <hannah@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/22206
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Yehuda Yitschak <yehuday@marvell.com>
---
 tools/marvell/debugger_scripts/XDB/spl_init.xdb | 39 ++++++++++++++++++++
 tools/marvell/debugger_scripts/XDB/startup.xdb  | 48 +++++++++++++++++++++++++
 2 files changed, 87 insertions(+)
 create mode 100644 tools/marvell/debugger_scripts/XDB/spl_init.xdb
 create mode 100644 tools/marvell/debugger_scripts/XDB/startup.xdb

diff --git a/tools/marvell/debugger_scripts/XDB/spl_init.xdb b/tools/marvell/debugger_scripts/XDB/spl_init.xdb
new file mode 100644
index 0000000..70f380e
--- /dev/null
+++ b/tools/marvell/debugger_scripts/XDB/spl_init.xdb
@@ -0,0 +1,39 @@
+! Set breakpoint to the first access to stack area and allow BootROM to run until it.
+! The memory in L2 cache should be ready when the execution terminated
+SET BREAKPOINT AT (@first_breakp) HARD
+RUN
+DELETE BREAKPOINT /ALL
+
+! Once the memory region in L2 cache is ready, load the BIN header and execute it
+PRINT /DBVAR=cmd "LOAD /SEGMENT /DEBUG /GLOBAL /INIT=\"\" OF \"./spl/u-boot-spl\""
+set val @ret_val = @exec(@cmd)
+IF @ret_val == -1 THEN
+	print "Error: Missing SPL file ./spl/u-boot-spl\n"
+	GOTO SPL_INIT_END
+END
+
+! Break at the end of BIN header execution before jumping back top BootROM
+SET BREAKPOINT AT (@second_breakp) HARD
+RUN
+DELETE BREAKPOINT /ALL
+
+! DRAM should be ready now.
+! Disable MMU before loading the u-boot image.
+IF (@strcmp(@SOC, "ARMv7") == 0) THEN
+	SET SYSREGISTER /"COPROCESSOR_1" ARM_CONTROL=0xC53078
+ELSE
+	IF (@strcmp(@SOC, "ARMv8") == 0) THEN
+		SET REG SCTLR_EL3=0xC50838
+	END
+END
+
+! For loading and execution of u-boot, uncomment the following lines:
+!PRINT /DBVAR=cmd "LOAD /SEGMENT /DEBUG /GLOBAL /INIT=\"\" OF \"./u-boot\""
+!set val @ret_val = @exec(@cmd)
+!IF @ret_val == -1 THEN
+!	print "Error: Can't find ./u-boot file\n"
+!	GOTO SPL_INIT_END
+!END
+!RUN
+
+SPL_INIT_END:
diff --git a/tools/marvell/debugger_scripts/XDB/startup.xdb b/tools/marvell/debugger_scripts/XDB/startup.xdb
new file mode 100644
index 0000000..bba7899
--- /dev/null
+++ b/tools/marvell/debugger_scripts/XDB/startup.xdb
@@ -0,0 +1,48 @@
+DEFINE SYMBOL /ADDRESS=0xF06F8240 device_id_a8k
+DEFINE SYMBOL /ADDRESS=0xD0018238 device_id_a3xx
+SET VAL @dev_id_3xx = -1
+SET VAL @dev_id_8k = -1
+SET VAL @first_breakp = 0x0
+SET VAL @second_breakp = 0x0
+
+! Check if working on ARMv8, if not set ARMv7
+SET OPTION /outputtrace=on
+SHOW CPU
+IF @strfind(@OUTPUTTRACE, "ARMV8") != -1 THEN
+	PRINT /DBVAR=SOC "ARMv8"
+ELSE
+	PRINT /DBVAR=SOC "ARMv7"
+END
+DEL outputtrace
+SET OPTION /outputtrace=off
+
+! Read the device ID from the relevant address
+IF (@strcmp(@SOC, "ARMv7") == 0) THEN
+	SET VAL @dev_id_3xx = (device_id_a3xx & 0xFFFF0000) >> 16
+ELSE
+	IF (@strcmp(@SOC, "ARMv8") == 0) THEN
+		SET VAL @dev_id_8k = (device_id_a8k & 0xFFFF0000) >> 16
+	END
+END
+
+SET VAL @dev_id = @dev_id_3xx
+IF @dev_id == 0x6820 || @dev_id == 0x6810 || @dev_id == 0x6828 || @dev_id == 0x6811 THEN
+	PRINT "%s" "Running on Armada 38x\n"
+	define macro /button static_ddr_init "batch \"NO STATIC DRAM\"\n"
+	SET VAL @first_breakp = 0x40000040
+	SET VAL @second_breakp = 0x400000BC
+ELSE
+	SET VAL @dev_id = @dev_id_8k
+	IF @dev_id == 0x0 || @dev_id == 0x800 || @dev_id == 0x10 || @dev_id == 0x810 THEN
+		PRINT "%s" "Running on APN 806\n"
+		define macro /button static_ddr_init "batch \"NO STATIC DRAM\"\n"
+		SET VAL @first_breakp = 0xFFE1C048
+		SET VAL @second_breakp = 0xFFE1C07C
+	ELSE
+		PRINT "Unknown device ID %x\n" @dev_id
+	END
+END
+
+define macro /button dynamic_ddr_init "batch \"./tools/marvell/debugger_scripts/XDB/spl_init.xdb\"\n"
+
+STARTUP_END:
\ No newline at end of file
-- 
1.9.1

